plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/StateMachine.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeApplications   #-}
+-- | State machine that manages credential tokens
+module Language.PlutusTx.Coordination.Contracts.Prism.StateMachine(
+    IDState(..)
+    , IDAction(..)
+    , UserCredential(..)
+    , scriptInstance
+    , machineClient
+    , mkMachineClient
+    ) where
+
+import           Data.Aeson                                                (FromJSON, ToJSON)
+import           Data.Hashable                                             (Hashable)
+import           GHC.Generics                                              (Generic)
+import           Language.Plutus.Contract.StateMachine                     (State (..), StateMachine (..),
+                                                                            StateMachineClient (..),
+                                                                            StateMachineInstance (..), Void)
+import qualified Language.Plutus.Contract.StateMachine                     as StateMachine
+import qualified Language.PlutusTx                                         as PlutusTx
+import           Language.PlutusTx.Coordination.Contracts.Prism.Credential (Credential (..), CredentialAuthority (..))
+import qualified Language.PlutusTx.Coordination.Contracts.Prism.Credential as Credential
+import           Language.PlutusTx.Prelude
+import qualified Ledger.Constraints                                        as Constraints
+import           Ledger.Constraints.TxConstraints                          (TxConstraints)
+import           Ledger.Crypto                                             (PubKeyHash)
+import qualified Ledger.Typed.Scripts                                      as Scripts
+import           Ledger.Value                                              (TokenName, Value)
+import qualified Prelude                                                   as Haskell
+
+data IDState =
+    Active -- ^ The credential is active and can be used in transactions
+    | Revoked -- ^ The credential has been revoked and can't be used anymore.
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+data IDAction =
+    PresentCredential -- ^ Present the credential in a transaction
+    | RevokeCredential -- ^ Revoke the credential
+    deriving stock (Generic, Haskell.Eq, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+-- | A 'Credential' issued to a user (public key address)
+data UserCredential =
+    UserCredential
+        { ucAddress    :: PubKeyHash
+        -- ^ Address of the credential holder
+        , ucCredential ::  Credential
+        -- ^ The credential
+        , ucToken      :: Value
+        -- ^ The 'Value' containing a token of the credential
+        -- (this needs to be included here because 'Credential.token'
+        -- is not available in on-chain code)
```

Comment:
This feels suspicious, need to think about it.

Reply:
It's a common issue, see for example the [`FutureAccounts`](https://github.com/input-output-hk/plutus/blob/master/plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs#L116) datatype from the futures contract